configcatclient/rolloutevaluator.py

```diff
@@ -48,174 +169,553 @@ def evaluate(self, key, user, default_value, default_variation_id, settings):  #
+        targeting_rules = setting_descriptor.get(TARGETING_RULES, [])
+        percentage_rule_attribute = setting_descriptor.get(PERCENTAGE_RULE_ATTRIBUTE)
 
-        user_has_invalid_type = user is not None and type(user) is not User
+        context = EvaluationContext(key, user, visited_keys)
+
+        user_has_invalid_type = context.user is not None and not isinstance(context.user, User)
         if user_has_invalid_type:
             self.log.warning('Cannot evaluate targeting rules and %% options for setting \'%s\' '
-                             '(User Object is not an instance of User type).',
+                             '(User Object is not an instance of User type). '
+                             'You should pass a User Object to the evaluation methods like `get_value()` '
+                             'in order to make targeting work properly. '
+                             'Read more: https://configcat.com/docs/advanced/user-object/',
                              key, event_id=4001)
-            user = None
+            # We set the user to None and won't log further missing user object warnings
+            context.user = None
+            context.is_missing_user_object_logged = True
+
+        try:
+            if log_builder and is_root_flag_evaluation:
+                log_builder.append("Evaluating '{}'".format(key))
+                log_builder.append(" for User '{}'".format(context.user) if context.user is not None else '')
+                log_builder.increase_indent()
+
+            # Evaluate targeting rules (logically connected by OR)
+            if log_builder and len(targeting_rules) > 0:
+                log_builder.new_line('Evaluating targeting rules and applying the first match if any:')
+            for targeting_rule in targeting_rules:
+                conditions = targeting_rule.get(CONDITIONS, [])
+
+                if len(conditions) > 0:
+                    served_value = targeting_rule.get(SERVED_VALUE)
+                    value = get_value(served_value) if served_value is not None else None
+
+                    # Evaluate targeting rule conditions (logically connected by AND)
+                    if self._evaluate_conditions(conditions, context, salt, config, log_builder, value):
+                        served_value = targeting_rule.get(SERVED_VALUE)
+                        if served_value is not None:
+                            variation_id = served_value.get(VARIATION_ID, default_variation_id)
+                            log_builder and is_root_flag_evaluation and log_builder.new_line("Returning '%s'." % value)
+                            return value, variation_id, targeting_rule, None, None
+                    else:
+                        continue
+
+                # Evaluate percentage options of the targeting rule
+                log_builder and log_builder.increase_indent()
+                percentage_options = targeting_rule.get(TARGETING_RULE_PERCENTAGE_OPTIONS, [])
```

Comment:
I'd keep it because the long variable names break the layout of the code. (By the way, without the logging, we could use the same code here. Unfortunately, this kind of rich logging breaks the code readability many places 😞)